### ISSUE TITLE
Low accuracy of Sth-sth v2 pretrained model.

### ISSUE THREAD

**Issue by jd730** [mit-han-lab/temporal-shift-module#74]
Hi,

I tried to test sth-sth v2 pretrained model (TSM ResNet50 | 8 * 2clip), but I got much lower accuracy than that in your paper and Readme. Should I use different arguments?

* Single Clip (59.1 on your paper)

Script:
```
python test_models.py somethingv2 \
   --test_segments=8 --batch_size=64 -j 24 --test_crops=3 --twice_sample   \
 --weights='pretrained/TSM_somethingv2_RGB_resnet50_shift8_blockres_avg_segment8_e45.pth'
```

Test Log:
```
somethingv2: 174 classes
=> shift: True, shift_div: 8, shift_place: blockres

    Initializing TSN with base model: resnet50.
    TSN Configurations:
        input_modality:     RGB
        num_segments:       8
        new_length:         1
        consensus_module:   avg
        dropout_ratio:      0.8
        img_feature_dim:    256

=> base model: resnet50
Adding temporal shift...
=> n_segment per stage: [8, 8, 8, 8]
=> Processing stage with 3 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 4 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 6 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 3 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
video number:24777
video 0 done, total 0/24777, average 0.926 sec/video, moving Prec@1 54.688 Prec@5 92.188
video 1280 done, total 1280/24777, average 0.050 sec/video, moving Prec@1 54.911 Prec@5 81.250
video 2560 done, total 2560/24777, average 0.029 sec/video, moving Prec@1 55.564 Prec@5 81.441
video 3840 done, total 3840/24777, average 0.022 sec/video, moving Prec@1 54.611 Prec@5 80.763
video 5120 done, total 5120/24777, average 0.018 sec/video, moving Prec@1 54.745 Prec@5 81.501
video 6400 done, total 6400/24777, average 0.016 sec/video, moving Prec@1 54.873 Prec@5 81.590
video 7680 done, total 7680/24777, average 0.014 sec/video, moving Prec@1 55.114 Prec@5 81.779
video 8960 done, total 8960/24777, average 0.013 sec/video, moving Prec@1 54.887 Prec@5 81.782
video 10240 done, total 10240/24777, average 0.012 sec/video, moving Prec@1 54.920 Prec@5 82.114
video 11520 done, total 11520/24777, average 0.012 sec/video, moving Prec@1 55.024 Prec@5 82.312
video 12800 done, total 12800/24777, average 0.011 sec/video, moving Prec@1 54.952 Prec@5 82.400
video 14080 done, total 14080/24777, average 0.011 sec/video, moving Prec@1 54.893 Prec@5 82.438
video 15360 done, total 15360/24777, average 0.010 sec/video, moving Prec@1 54.895 Prec@5 82.346
video 16640 done, total 16640/24777, average 0.010 sec/video, moving Prec@1 54.795 Prec@5 82.280
video 17920 done, total 17920/24777, average 0.010 sec/video, moving Prec@1 54.621 Prec@5 82.184
video 19200 done, total 19200/24777, average 0.010 sec/video, moving Prec@1 54.620 Prec@5 82.107
video 20480 done, total 20480/24777, average 0.010 sec/video, moving Prec@1 54.644 Prec@5 82.248
video 21760 done, total 21760/24777, average 0.009 sec/video, moving Prec@1 54.683 Prec@5 82.299
video 23040 done, total 23040/24777, average 0.009 sec/video, moving Prec@1 54.748 Prec@5 82.358
video 24320 done, total 24320/24777, average 0.009 sec/video, moving Prec@1 54.724 Prec@5 82.271
[0.90517241 0.34868421 0.22222222 0.60194175 0.41025641 0.61842105
 0.74580336 0.54545455 0.5511811  0.57615894 0.63483146 0.66423358
 0.29059829 0.35       0.71578947 0.50212766 0.43654822 0.32236842
 0.29288703 0.40758294 0.31889764 0.57017544 0.30813953 0.46808511
 0.43181818 0.42105263 0.5        0.54455446 0.8        0.55938697
 0.75675676 0.69369369 0.89772727 0.15533981 0.1686747  0.
 0.79723502 0.71863118 0.2037037  0.52941176 0.68306011 0.67857143
 0.57276995 0.65594855 0.74324324 0.57938719 0.58433735 0.33165829
 0.37888199 0.75698925 0.67932489 0.         0.15625    0.48571429
 0.16666667 0.52666667 0.43835616 0.59433962 0.17647059 0.71223022
 0.68       0.23529412 0.50632911 0.14018692 0.05405405 0.48584906
 0.20754717 0.43283582 0.66666667 0.17       0.5        0.39130435
 0.51470588 0.61538462 0.68867925 0.23684211 0.38888889 0.42307692
 0.22222222 0.63793103 0.109375   0.096      0.17021277 0.31914894
 0.69333333 0.53968254 0.76470588 0.8        0.16666667 0.10843373
 0.44927536 0.7        0.42105263 0.612      0.76271186 0.13888889
 0.17647059 0.19834711 0.63265306 0.73626374 0.42641509 0.58791209
 0.21551724 0.5625     0.65354331 0.72592593 0.42808219 0.57142857
 0.28378378 0.50333333 0.43661972 0.         0.4028777  0.33333333
 0.64236111 0.41304348 0.48148148 0.6984127  0.30508475 0.58415842
 0.73228346 0.67045455 0.408867   0.55384615 0.63461538 0.52752294
 0.39805825 0.33043478 0.3256705  0.55147059 0.22142857 0.15068493
 0.43678161 0.14035088 0.78532609 0.47852761 0.11764706 0.46666667
 0.25688073 0.64593301 0.63565891 0.203125   0.5625     0.54220779
 0.48407643 0.45592705 0.75277778 0.0703125  0.48535565 0.89245283
 0.69565217 0.3423913  0.29591837 0.79816514 0.53543307 0.18292683
 0.39230769 0.58695652 0.52752294 0.23636364 0.23853211 0.36842105
 0.53271028 0.425      0.81585678 0.85483871 0.93571429 0.91358025
 0.88333333 0.76       0.42647059 0.69551282 0.78139535 0.68613139]
upper bound: 0.5014762756861307
-----Evaluation is finished------
Class Accuracy 48.14%
Overall Prec@1 54.72% Prec@5 82.29%
```

* 2 Clips (61.2 according to Readme)
Script:
```
python test_models.py somethingv2 \
   --test_segments=8 --batch_size=64 -j 24 --test_crops=3 --twice_sample   \
 --weights='pretrained/TSM_somethingv2_RGB_resnet50_shift8_blockres_avg_segment8_e45.pth'
```

Test Log:
```
somethingv2: 174 classes
=> shift: True, shift_div: 8, shift_place: blockres

    Initializing TSN with base model: resnet50.
    TSN Configurations:
        input_modality:     RGB
        num_segments:       8
        new_length:         1
        consensus_module:   avg
        dropout_ratio:      0.8
        img_feature_dim:    256

=> base model: resnet50
Adding temporal shift...
=> n_segment per stage: [8, 8, 8, 8]
=> Processing stage with 3 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 4 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 6 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Processing stage with 3 blocks residual
=> Using fold div: 8
=> Using fold div: 8
=> Using fold div: 8
=> Using twice sample for the dataset...
video number:24777
video 0 done, total 0/24777, average 1.546 sec/video, moving Prec@1 59.375 Prec@5 92.188
video 1280 done, total 1280/24777, average 0.112 sec/video, moving Prec@1 56.324 Prec@5 83.259
video 2560 done, total 2560/24777, average 0.075 sec/video, moving Prec@1 57.127 Prec@5 83.575
video 3840 done, total 3840/24777, average 0.060 sec/video, moving Prec@1 56.276 Prec@5 82.889
video 5120 done, total 5120/24777, average 0.052 sec/video, moving Prec@1 56.674 Prec@5 83.314
video 6400 done, total 6400/24777, average 0.047 sec/video, moving Prec@1 56.668 Prec@5 83.323
video 7680 done, total 7680/24777, average 0.044 sec/video, moving Prec@1 56.792 Prec@5 83.355
video 8960 done, total 8960/24777, average 0.042 sec/video, moving Prec@1 56.549 Prec@5 83.344
video 10240 done, total 10240/24777, average 0.040 sec/video, moving Prec@1 56.677 Prec@5 83.667
video 11520 done, total 11520/24777, average 0.039 sec/video, moving Prec@1 56.733 Prec@5 83.943
video 12800 done, total 12800/24777, average 0.037 sec/video, moving Prec@1 56.678 Prec@5 84.002
video 14080 done, total 14080/24777, average 0.036 sec/video, moving Prec@1 56.681 Prec@5 84.036
video 15360 done, total 15360/24777, average 0.036 sec/video, moving Prec@1 56.717 Prec@5 84.012
video 16640 done, total 16640/24777, average 0.035 sec/video, moving Prec@1 56.597 Prec@5 83.956
video 17920 done, total 17920/24777, average 0.035 sec/video, moving Prec@1 56.456 Prec@5 83.880
video 19200 done, total 19200/24777, average 0.034 sec/video, moving Prec@1 56.458 Prec@5 83.783
video 20480 done, total 20480/24777, average 0.034 sec/video, moving Prec@1 56.450 Prec@5 83.883
video 21760 done, total 21760/24777, average 0.034 sec/video, moving Prec@1 56.502 Prec@5 83.967
video 23040 done, total 23040/24777, average 0.033 sec/video, moving Prec@1 56.579 Prec@5 83.985
video 24320 done, total 24320/24777, average 0.033 sec/video, moving Prec@1 56.574 Prec@5 83.871
[0.94827586 0.34868421 0.26143791 0.63106796 0.41025641 0.63157895
 0.7793765  0.57575758 0.55905512 0.62251656 0.61797753 0.70072993
 0.32905983 0.35       0.76140351 0.50212766 0.41624365 0.32236842
 0.33891213 0.38388626 0.33464567 0.60526316 0.3255814  0.5106383
 0.52272727 0.42105263 0.475      0.56435644 0.78181818 0.57471264
 0.77027027 0.7027027  0.90909091 0.16504854 0.21686747 0.
 0.8156682  0.74524715 0.18518519 0.5        0.66666667 0.71428571
 0.58685446 0.64630225 0.72972973 0.59052925 0.58433735 0.40201005
 0.41614907 0.7827957  0.71308017 0.         0.1875     0.57142857
 0.125      0.57333333 0.45205479 0.6509434  0.23529412 0.72661871
 0.68       0.23529412 0.5443038  0.11214953 0.10810811 0.53773585
 0.26415094 0.44776119 0.69298246 0.16       0.53571429 0.47826087
 0.52941176 0.64102564 0.66037736 0.31578947 0.33333333 0.38461538
 0.22222222 0.67241379 0.140625   0.088      0.19148936 0.31914894
 0.69333333 0.52380952 0.77058824 0.792      0.25       0.10843373
 0.44927536 0.71666667 0.42105263 0.66       0.77542373 0.19444444
 0.17647059 0.18181818 0.62755102 0.73626374 0.4490566  0.61538462
 0.25862069 0.59821429 0.71653543 0.75555556 0.45890411 0.5862069
 0.32432432 0.50666667 0.46478873 0.         0.41726619 0.36111111
 0.6875     0.45652174 0.48148148 0.76984127 0.40677966 0.57425743
 0.77165354 0.67045455 0.38916256 0.52307692 0.63461538 0.53211009
 0.39805825 0.32173913 0.32183908 0.58088235 0.22857143 0.16438356
 0.55172414 0.14035088 0.7798913  0.51840491 0.17647059 0.53333333
 0.29357798 0.66028708 0.65116279 0.28125    0.546875   0.55844156
 0.49044586 0.47112462 0.75833333 0.078125   0.50627615 0.89056604
 0.7201087  0.39673913 0.35714286 0.8440367  0.56692913 0.17073171
 0.4        0.5923913  0.56422018 0.23636364 0.2293578  0.44736842
 0.53271028 0.525      0.85933504 0.88709677 0.94285714 0.9382716
 0.88333333 0.82666667 0.41666667 0.70192308 0.8        0.70072993]
upper bound: 0.5229737739783421
-----Evaluation is finished------
Class Accuracy 50.11%
Overall Prec@1 56.58% Prec@5 83.91%
```

Thank you.

**Comment by jd730** [comment]
The score of 2 Clip seems similar to Table 2 on http://openaccess.thecvf.com/content_ICCV_2019/papers/Xiao_Reasoning_About_Human-Object_Interactions_Through_Dual_Attention_Networks_ICCV_2019_paper.pdf
![image](https://user-images.githubusercontent.com/9247850/79038418-0f884380-7c14-11ea-93c3-dc9c39dbaf24.png)

**Comment by tonylins** [comment]
Hi, for the results in Dual Attention paper, actually they reference an early version of the results (v1 arXiv), where we did not set learning rate right. After that, with an increased learning rate, the results are much higher. 

Regarding the performance of the pre-trained model, how did you pre-process the data?

**Comment by jd730** [comment]
@tonylins  Thank you for reply. Since Sth sth v2 has official image frames, I used them with labels converted by `gen_label_sthv2.py` when I ran above mentioned script.

**Comment by tonylins** [comment]
Hi, I don't think Sth V2 has official frames. Where did you get them?

**Comment by jd730** [comment]
I am sorry. I was confused with Sth Sth v1 which has jpg images (https://20bn.com/datasets/something-something/v1).

The official file format of Sth Sth v2 is webm using VP9 as encoding.
I used [`vid2img_sthv2.py`](https://github.com/mit-han-lab/temporal-shift-module/blob/master/tools/vid2img_sthv2.py) on your repositories.

**Comment by Andy1621** [comment]
> I am sorry. I was confused with Sth Sth v1 which has jpg images (https://20bn.com/datasets/something-something/v1).
> 
> The official file format of Sth Sth v2 is webm using VP9 as encoding.
> I used [`vid2img_sthv2.py`](https://github.com/mit-han-lab/temporal-shift-module/blob/master/tools/vid2img_sthv2.py) on your repositories.
> As a result I got images from a video something like [this](https://drive.google.com/open?id=1D0pGk0mr99iaADHs4BKxW-pWZd3g_kkd)

I also got the same lower accuracy in sth-sth v1 and kinetics #65 , and I tried to fine-tuing the pre-trained model and the same accuracy. (lr=0.005, epoch=30, step decay => divided by 10 in 10 20 25 epoch)

**Comment by jd730** [comment]
@Andy1621  Thank you for sharing your experience. It is strange that more training from an already trained model is needed to achieve the reported scores.
@tonylins, I found that you originally (arXiv v1) fine-tune your model from Kinectics pretrained one for 45 epochs. 

> we followed the common practice [43, 53, 3, 46] to first pre-trained the models on Kinetics [21] for 45 epochs with initial learning rate 0.01 and decayed by 0.1 every 15 epochs. We then fine-tuned the model to other target datasets like SomethingSomething [12], UCF101 [34], andHMDB51 [22]. The fine-tuning is conducted for 25 epochs with initial learning rate 0.001 and decayed by a factor of 0.1 every 10 epochs. We follow the practice in [42] to fix all the Batch Normalization [19] layers except for the first one.

Moreover, checkpoint link did not change since it was on the public. Is the checkpoint related to the old version? If so, it is still strange since it should achieve higher than 58.

Thanks.

**Comment by tonylins** [comment]
Hi @jd730, all the models are for the new version. I report the accuracy in the README. I also attach test log of the model. It should reach the reported performance if all the environments are the same. 
But right now I am not sure what is different regarding the environment set up.

**Comment by tonylins** [comment]
I recommend training a model using the same hyper-parameter on your own dataset set up, which should give the reported accuracy.

**Comment by jd730** [comment]
What is the right parameter?
According to #51, Is the learning rate for sth v2 0.025, not 0.01?

Currently, I am running with using below script.
```
python main.py something RGB
--arch resnet50 --num_segments 8
--gd 20 --lr 0.01 --wd 1e-4 --lr_steps 20 40 --epochs 50
--batch-size 64 -j 32 --dropout 0.5 --consensus_type=avg --eval-freq=1
--shift --shift_div=8 --shift_place=blockres --npb
```

I will let you know once it is done.


When I used 
```
python main.py somethingv2 RGB \
     --arch resnet50 --num_segments 16 \
     --gd 20 --lr 0.001 --lr_steps 10 20 --epochs 25 \
     --batch-size 64 -j 16 --dropout 0.5 --consensus_type=avg --eval-freq=1 \
     --shift --shift_div=8 --shift_place=blockres \
     --tune_from=pretrained/TSM_kinetics_RGB_resnet50_shift8_blockres_avg_segment16_e50.pth
```
which is introduced on ReadMe, it achieves 55.1 as test_top1_best during training. I think this script is outdated as you mentioned on other issues.

**Comment by tonylins** [comment]
Hi @jd730, the first script should give a reasonable performance. Let me know if the results are right!

**Comment by jd730** [comment]
Thank @tonylins, I will let you know in 36 hours. Currently, at 11 epochs, acc/test_top1_best is 52.

**Comment by tonylins** [comment]
@jd730 Sure! And thanks for the reminder, I forgot to modify the training script in README until now.

**Comment by jd730** [comment]
Hi @tonylins.
My job got 58.595 at 25 epoch and then slightly decreasing.

![image](https://user-images.githubusercontent.com/9247850/79197704-2bd2ed00-7e6d-11ea-9242-32a8caea4dd5.png)

**Comment by Sobeney** [comment]
@jd730 Hi, I have faced the same problem when test the given pretrained model on SSV2 (8framex2clip one), the result when using twice sample is around 57. I just wonder have you solved the problem? Can you get the reported result (61.1) by using the given pretrained model? Or is that you can only get considerable result (60.0) by your self-trained model? 

Thanks a lot and looking forward for your reply.

**Comment by jd730** [comment]
Hi @Sobeney, I am so sorry that I cannot answer. I forgot whether I solved the result or not. The only thing I remember is that  I extracted frames again after writing the last comment (not sure it was a solution).

**Comment by Sobeney** [comment]
> Hi @Sobeney, I am so sorry that I cannot answer. I forgot whether I solved the result or not. The only thing I remember is that I extracted frames again after writing the last comment (not sure it was a solution).

Thanks a lot! I also doult the data problem... By the way, I wonder what script did you used to re-extract the frames? Following this project's guide (vid2img_sthv2.py)?

**Comment by jd730** [comment]
I guess so. Or, [mmaction](https://github.com/open-mmlab/mmaction) repo.
